src/charm.py

```diff
@@ -184,6 +178,37 @@ def main(self, event):
 
         self.model.unit.status = ActiveStatus()
 
+    def _send_info(self, event, interfaces):
+        if interfaces["kubeflow-profiles"]:
+            interfaces["kubeflow-profiles"].send_data(
+                {
+                    "service-name": self.model.app.name,
+                    "service-port": str(self.model.config["port"]),
+                }
+            )
+
+    def _check_leader(self):
+        if not self.unit.is_leader():
+            # We can't do anything useful when not the leader, so do nothing.
+            raise CheckFailed("Waiting for leadership", WaitingStatus)
+
+    def _get_interfaces(self):
+        try:
+            interfaces = get_interfaces(self)
+        except NoVersionsListed as err:
+            raise CheckFailed(err, WaitingStatus)
+        except NoCompatibleVersions as err:
+            raise CheckFailed(err, BlockedStatus)
+        return interfaces
```

Comment:
We should move this entire pattern into SDI.  I dont think there's a reason we can't make this the way it naturally emits errors.  If the `Status` attributes were on these exceptions, we could add them to this block in main:
```python
except CheckFailed, NoVersionsListed, NoCompatibleVersions as e:
            self.model.unit.status = e.status
```

Reply:
The `OCIImageResourceError` provides status, but I don't necessarily like that. It abstracts away the status setting into a library which isn't as intuitive. That is what I like about raising `CheckFailed` and provideing the status there, it shows the status you are going to end up raising in the charm code.

Reply:
It does add abstraction, but I don't know if it is bad.  It is the same pattern as used with the `OCIImageResourceError` in most of our charms:
```python
try:
    profile_image_details = self.profile_image.fetch()
    kfam_image_details = self.kfam_image.fetch()
except OCIImageResourceError as e:
    self.model.unit.status = e.status
    self.log.info(e)
```

We control it (we are the ones that maintain the SDI library) and we have tests here that assert the outcome we expect (so if the SDI changed NoVersionsListed from `Waiting` to `Active` we catch it here), so it feels safe and convenient to me

Regardless though, we shouldn't hold this PR up on changing the SDI.

Reply:
Yes, and as I mentioned, I don't really like that pattern, a library shouldn't be the one deciding the status of the charm, that should be up to the charm author.

One of the number one priorities in charm code should always be readability before convenience. This is operations code, and operators will want to know exactly what is happening with their application. This is why I believe we have to be very careful with abstracting away certain functionalities just to make our lives easier.